infra/src/main/kotlin/fi/fta/geoviite/infra/linking/SwitchLinkingService.kt

```diff
@@ -1021,6 +1022,17 @@ class SwitchLinkingService @Autowired constructor(
         )
     }
 
+    @Transactional(readOnly = true)
+    fun getSuggestedSwitchesAtPresentationJointLocations(switches: List<TrackLayoutSwitch>): List<Pair<DomainId<TrackLayoutSwitch>, SuggestedSwitch?>> {
+        logger.serviceCall("getSuggestedSwitchesAtPresentationJointLocations", "switches" to switches)
+        return switches.parallelStream().map { switch ->
```

Comment:
Täällä huomionarvoisena asiana säikeistys

Reply:
Mmh, vähän sama juttu kuin siinä toisessa rinnakkaistus-optimoinnissa: näitä säikeitä ei kannata ampua yli servicerajojen ja kannassa käynnit olisi syytä hoitaa yhdessä ja samassa, suoraan controllerissa tulleessa säikeessä. Syynä siis käyttäjätietojen kuljetus thread contextissa, connection poolin toiminta, transaktioiden toiminta jne. Perusarkkitehtuurina meillä on 1-säie-per-kutsu, ja se on hyvä malli koska se pitää asiat koodin kannalta simppelinä. Silloin kuin rajumpi laskenta vaatii rinnakkaistusta, kannattaa ensin hakea kaikki data (mieluiten könttä-hakuna) ja rinnakkaistaa vasta sen rouskutus (puhdas funktio). Silloin monisäikeisyyden vaikutukset rajautuu tiettyyn käyttöpaikkaan ja on helpommin hallittavissa.

Reply:
Käytin tässä aamulla jonkin verran aikaa siihen että koitin tutkia ja parannella tätä. Selkeästi suurin osa ajasta kuluu vaihde-ehdotuksen laskemiseen (ei kantahakuihin vaan ihan laskentaan.) Koitin parhaani mukaan isoloida laskennan erilleen kantahauista ja rinnakkaistaa pelkästään sen. Tällöin rinnakkaistuksesta saadaan edelleen about täydet hyödyt, mutta kannassa käydään vaan yhdestä threadista. Ajoajat ovat käytännössä mulla lokaalisti raiteelle 006 IMA-KON (koko kutsulle, ei pelkästään vaihde-ehdotuslaskennalle) 4.5sec rinnakkaistettuna vs. 11-12sec yhdessä säikeessä, joten sikäli rinnakkaistuksesta on kyllä saatavissa merkittäviä hyötyjä.

Käydään silti vaan tämä läpi yhdessä dailyn yhteydessä. Arihan oli tainnut päätyä poistamaan rinnakkaistusta omilla tiketeillään, joten voisi puhua mennäänkö tässä edelleen tällä parannetulla rinnakkaistetulla versiolla, vai halutaanko täällä tehdä samoin (lisäähän se kuitenkin monimutkaisuutta, tuotannonkin pannut ovat mitä ovat coremääriltään, tämä kutsu on tarkoitus ampua napinpainalluksesta, ja rinnakkaistettunakin tätä joutuu odottamaan runsasvaihteisilla raiteille)